### PR TITLE
[Dy2Stat & Quantization]Support append customize attributes into op_desc in nn.Layer

### PR DESCRIPTION
**Pull request by Aurelius84**
### PR types

New features
### PR changes

APIs
#### 1. PR 描述
Support append customize attributes into op_desc in nn.Layer

+ 支持控制流 sub block
+ 支持同一个layer多次调用

此PR是基于layer的hook实现的，经与@军才沟通，量化侧所有的普通API （如paddle.sub）等都会封装成nn.layer对象，因此都可以基于此PR实现op attribute的指定与保存

image

### CODE REVIEW

File: python/paddle/fluid/dygraph/layers.py [hunk 1]
```diff
@@ -529,9 +534,8 @@ def parameters(self, include_sublayers=True):
 
         """
         ret = [
-            param
-            for _, param in self.named_parameters(
-                include_sublayers=include_sublayers)
+            param for _, param in
+            self.named_parameters(include_sublayers=include_sublayers)
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/layers.py:538]: just code format

File: python/paddle/fluid/dygraph/layers.py [hunk 2]
```diff
@@ -786,9 +790,8 @@ def buffers(self, include_sublayers=True):
 
         """
         ret = [
-            buffer
-            for _, buffer in self.named_buffers(
-                include_sublayers=include_sublayers)
+            buffer for _, buffer in
+            self.named_buffers(include_sublayers=include_sublayers)
```
Comment (Aurelius84) [inline python/paddle/fluid/dygraph/layers.py:794]: just code format

File: python/paddle/fluid/dygraph/layer_hooks.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        self.hooks = hooks
+
+
+def record_program_ops_pre_hook(layer, inputs):
+    """
+    A pre-hook to mark op numbers before enter layer.forward.
+    """
+    if not in_dygraph_mode():
+        if layer._op_recorder.start < 0:
+            layer._op_recorder.start = len(default_main_program().block(0).ops)
```
Comment (juncaipeng) [inline python/paddle/fluid/dygraph/layer_hooks.py]: If the model has multiple blocks, can the attribute be saved to the right op?

File: python/paddle/fluid/dygraph/layers.py [hunk 4]
```diff
@@ -1028,6 +1033,35 @@ def forward(self, input):
...
+            attrs(dict): customized attributes that will be added into op_descs.
+
+        NOTE: The interface is only exposed to developers.
+        """
+        if not isinstance(attrs, dict):
+            raise TypeError("attrs should be type(dict), but received {}".
+                            format(type(attrs).__name__))
+
+        if len(self._customized_attrs) != 0:
+            raise RuntimeError("Unsupported to call self._set_op_attrs twice!")
```
Comment (juncaipeng) [inline python/paddle/fluid/dygraph/layers.py]: Can this restriction be removed？ The new attributes can overwrite old attributes. 
Maybe it should analyze whether the hooks have been registered

File: python/paddle/fluid/dygraph/layer_hooks.py (hunk 3 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/layer_hooks.py]: Yes， I support set attribute in sub block

File: python/paddle/fluid/dygraph/layers.py (hunk 4 above)
Comment (Aurelius84) [reply python/paddle/fluid/dygraph/layers.py]: Removed it.

**Review by juncaipeng — APPROVED** [review]
LGTM

**Pull request merged**